### PR TITLE
ddl: fix adding multi-value index

### LINKED ISSUE #51162: data and index are inconsistent after adding multi-value index
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

``` SQL
CREATE TABLE `tl` (
  `col_42` json NOT NULL,
  `col_43` tinyint(1) DEFAULT NULL,
  `col_44` char(168) CHARACTER SET gbk COLLATE gbk_bin DEFAULT NULL,
  `col_45` json DEFAULT NULL,
  `col_46` text COLLATE utf8mb4_unicode_ci NOT NULL,
  `col_47` char(43) COLLATE utf8mb4_unicode_ci NOT NULL DEFAULT 'xW2YNb99pse4)',
  `col_48` time NOT NULL DEFAULT '12:31:25',
  PRIMARY KEY (`col_47`,`col_46`(2)) /*T![clustered_index] CLUSTERED */
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;

INSERT INTO `tl` VALUES ('[\"01qHMy8yycbq2v1CZ80kiIECEQFLXYsnM0tpVh9vbx71DMpaPdp6hBkrpTodkXVt\", \"dGugZHV3pmqjLSdICAYbiwFGZsYWX9MrzgSGCslmEsfFzXYOiXyE0YMS1TSfnX5R\", \"Ypz8Jj0hBSBQbiWd3LbPRV2qvSEAxV8rlXnxpn93nXmFyMJkCQzyhtNn0FUwH6Ce\", \"Yf6kag9ruOT04V8AHT6NYhCNIa2CYZDVg8zWaxxHFm0GPAJh0e6Mc5Upa0EAp1Sr\", \"ljrRnPUdxJNkWxmkn4K25YeFwVtedGtLDrWlWYZyqt2Shpfg4p7jroOwbOZHuHiE\"]',0,'釸7鈡w歀痂暰苕rv)Ts','[5661899762065804824, 3294089805449875423, 2378579521053711229, 6262068190112960080]','Wxup81婱鍋cH爀
2潺檛楍1N7','2ZNeu2n5-2+Q3qYp','10:14:20'), ('[\"iOl0B9dVMoNYk8Qroj6JTwm0MaCJmr5qnQzdjMtaoDar4jGe6q20HAyVu3GacogW\", \"PBy3AXmGhzJMBX3Ow9INdeATLld8bVVjrIUVTlspzpHMZPuFGlEsDeRN6JOLqRLy\", \"jyr0odf7TOhLpuejrfbvsBbjqCchHSKEFZ079c5KUhu0FIuoqm6tBNbWCBb0xb3c\", \"PmPE8oZf73Zwffi97D7AaQM5g0xameUwPvO15LEUPVlF4iiDVsXyGMESSbEzR5L7\", \"4y0WxQIzPw217QeSXCySsK3pGbHdMxx6ETtIZQfGI8v7Pq9Rxhwkslr1ih8M7Dok\"]',0,'','[2142591371645505560, 2975660707908269211]','','aC794Ir-)iJZr&a','02:30:55'), ('[\"je9igPre8L6VdmtEtVDwxCgTqcuSKdqdO4fheHXRJOQLojiHx4F8cXmFpiYq9955\", \"zPjn7sAeMVLLGE3o9lpnZai26ad8qxHNQF6CSsMshFbP2fDuByV3SYxhxkVwIQd3\", \"Etj8PQZe64XWTbBElIEhWA3F3OcytxNMLvxQqr4miDxoK96Q8fGjBg34H3i4y3V5\", \"IDNAQjueBRQyUE285zoPSMhIvVKrxbiXAVvge0iC9AaEFK18Gh3sign84Nu6M3Ar\", \"MK79tEDnyHn2Cmw4W55hEzAsKvqUTZbViv6UyiEqnUE9jpBOnMEDQbq6yQip4qm3\"]',1,')譺
)RK5','[1760220588316922379, 4921535278728445755, 5623214347361271244, 5272407040485486430, 9017432485293265763]','RLFF','gS','06:27:28');

alter table tl add index idx_16(`col_48`,(cast(`col_45` as signed array)),`col_46`(5));

admin check table tl;
```

### 2. What did you expect to see? (Required)
No error
### 3. What did you see instead (Required)
``` sql
mysql> admin check table tl;
ERROR 8134 (HY000): data inconsistency in table: tl, index: idx_16, col: col_46, handle: "{\x0e+\x10j\x0fd\x0e\x8b\x10\x1f\x0e+\x0fd\x0e.\x02!\x0e+\x04(\x0f\xb4\x0e,\x0f\xb4\x10^\x0f\xa7, \x10Q\x10Z}", index-values:"KindString Wx" != record-values:"KindString Wxup8", compare err:<nil>
```
### 4. What is your TiDB version? (Required)
4e416993cc65b4da6786a7ce859b2ab99e5bba15

### PR DESCRIPTION
**Pull request by ywqzzy**
### What problem does this PR solve?

Issue Number: close #51162 

Problem Summary:

### What changed and how does it work?
Restored data should not truncated in buildHandle.
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/ddl/index.go [hunk 1]
```diff
@@ -1825,7 +1827,11 @@ func writeChunkToLocal(
 			idxDataBuf = extractDatumByOffsets(
 				row, copCtx.IndexColumnOutputOffsets(idxID), c.ExprColumnInfos, idxDataBuf)
 			idxData := idxDataBuf[:len(index.Meta().Columns)]
-			rsData := getRestoreData(c.TableInfo, copCtx.IndexInfo(idxID), c.PrimaryKeyInfo, handleDataBuf)
+			var rsData []types.Datum
+			if tables.NeedRestoredData(index.Meta().Columns, c.TableInfo.Columns) {
+				restoreDataBuf := extractDatumByOffsets(row, c.HandleOutputOffsets, c.ExprColumnInfos, restoreDataBuf)
```
Comment (tangenta) [inline pkg/ddl/index.go:1832]: I think cloning a `handleDataBuf` is more effective.

File: pkg/ddl/index.go [hunk 2]
```diff
@@ -1801,6 +1801,8 @@ func writeChunkToLocal(
 	maxIdxColCnt := maxIndexColumnCount(indexes)
 	idxDataBuf := make([]types.Datum, maxIdxColCnt)
 	handleDataBuf := make([]types.Datum, len(c.HandleOutputOffsets))
+	restoreDataBuf := make([]types.Datum, len(c.HandleOutputOffsets))
```
Comment (tangenta) [inline pkg/ddl/index.go:1804]: Lazy initialization is better.

File: pkg/ddl/index.go [hunk 3]
```diff
@@ -1816,6 +1818,20 @@ func writeChunkToLocal(
...
 		handleDataBuf := extractDatumByOffsets(row, c.HandleOutputOffsets, c.ExprColumnInfos, handleDataBuf)
+		needRestoreForIndexes := make([]bool, len(indexes))
+		restore := false
+		for i, index := range indexes {
+			needRestore := tables.NeedRestoredData(index.Meta().Columns, c.TableInfo.Columns)
+			needRestoreForIndexes[i] = needRestore
+			restore = restore || needRestore
+		}
+		if restore {
+			restoreDataBuf = make([]types.Datum, len(c.HandleOutputOffsets))
```
Comment (tangenta) [inline pkg/ddl/index.go:1829]:
```suggestion
			if restoreDataBuf == nil {
				restoreDataBuf = make([]types.Datum, len(c.HandleOutputOffsets))
			}
```

**Review by tangenta — COMMENTED** [review]
Rest LGTM

**Review by tangenta — APPROVED** [review]

File: tests/realtikvtest/addindextest/add_index_test.go [hunk 4]
```diff
@@ -112,3 +112,29 @@ func TestAddForeignKeyWithAutoCreateIndex(t *testing.T) {
...
+	 col_44 char(168) CHARACTER SET gbk COLLATE gbk_bin DEFAULT NULL,
+	 col_45 json DEFAULT NULL,
+	 col_46 text COLLATE utf8mb4_unicode_ci NOT NULL,
+	 col_47 char(43) COLLATE utf8mb4_unicode_ci NOT NULL DEFAULT 'xW2YNb99pse4)',
+	 col_48 time NOT NULL DEFAULT '12:31:25',
+	 PRIMARY KEY (col_47,col_46(2)) /*T![clustered_index] CLUSTERED */
+	  ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;`)
+
+	tk.MustExec(`INSERT INTO tl VALUES
+	('[\"01qHMy8yycbq2v1CZ80kiIECEQFLXYsnM0tpVh9vbx71DMpaPdp6hBkrpTodkXVt\", \"dGugZHV3pmqjLSdICAYbiwFGZsYWX9MrzgSGCslmEsfFzXYOiXyE0YMS1TSfnX5R\", \"Ypz8Jj0hBSBQbiWd3LbPRV2qvSEAxV8rlXnxpn93nXmFyMJkCQzyhtNn0FUwH6Ce\", \"Yf6kag9ruOT04V8AHT6NYhCNIa2CYZDVg8zWaxxHFm0GPAJh0e6Mc5Upa0EAp1Sr\", \"ljrRnPUdxJNkWxmkn4K25YeFwVtedGtLDrWlWYZyqt2Shpfg4p7jroOwbOZHuHiE\"]',0,'釸7鈡w歀痂暰苕rv)Ts','[5661899762065804824, 3294089805449875423, 2378579521053711229, 6262068190112960080]','Wxup81婱鍋cH爀
```
Comment (wjhuang2016) [inline tests/realtikvtest/addindextest/add_index_test.go:133]: Please simplify the values.

Comment (ywqzzy) [reply tests/realtikvtest/addindextest/add_index_test.go:133]: ok

**Review by Benjamin2037 — APPROVED** [review]
LGTM

**Pull request merged**